File: apps/api/src/sample/sample.service.ts
import { Injectable } from '@nestjs/common';
import { S3Service } from '@/lib/s3.service';
import { SAMPLE_DIR } from '@workspace/shared';
import configuration from '@/config/configuration';
import { SqsService } from '@/lib/sqs.service';
import { R2Service } from '@/lib/r2.service';
import { DatabaseService } from '@/db/database.service';
import { domainEvents } from '@workspace/db';

interface uploadImageProps {
  image: string;
  extension: string;
}

interface sendMessageProps {
  message: string;
}

@Injectable()
export class SampleService {
  constructor(
    private readonly s3Service: S3Service,
    private readonly sqsService: SqsService,
    private readonly databaseService: DatabaseService,
    private readonly r2Service: R2Service,
  ) {}

  async uploadImage(input: uploadImageProps, actorId: string) {
    const { image, extension } = input;
    const base64Data = image.replace(/^data:image\/\w+;base64,/, '');
    const buffer = Buffer.from(base64Data, 'base64');
    const timestamp = Date.now();
    const fileName = `${SAMPLE_DIR}/placeholder_${timestamp}.${extension}`;

    try {
      await this.s3Service.uploadImage(
        fileName,
        buffer,
        `image/${extension === 'jpg' ? 'jpeg' : extension}`,
      );

      const url = `${this.r2Service.r2PublicUrl}/${fileName}`;

      await this.databaseService.db.insert(domainEvents).values({
        eventType: 'upload_sample_image_r2',
        actorId,
        payload: { success: true, error: '', url },
      });

      return { success: true, url };
    } catch (error) {
      console.error('Sample image upload failed:', error);

      await this.databaseService.db.insert(domainEvents).values({
        eventType: 'upload_sample_image_r2',
        actorId,
        payload: {
          success: false,
          error: error instanceof Error ? error.message : String(error),
          url: '',
        },
      });

      throw error;
    }
  }

  async sendMessage(input: sendMessageProps, actorId: string) {
    try {
      const res = await this.sqsService.sendMessage(
        configuration().sampleSqsQueueUrl,
        { message: input.message },
      );

      if (!res.ok) {
        throw new Error('Failed to send sample message');
      }

      const messageId = res.MessageId ?? '';
      const message = input.message + ' Acknowledged!';

      await this.databaseService.db.insert(domainEvents).values({
        eventType: 'send_sample_sqs_message',
        actorId,
        payload: { success: true, error: '', messageId, message },
      });

      return { success: true, messageId, message };
    } catch (error) {
      console.error('Sending sample message failed:', error);

      await this.databaseService.db.insert(domainEvents).values({
        eventType: 'send_sample_sqs_message',
        actorId,
        payload: {
          success: false,
          error: error instanceof Error ? error.message : String(error),
          messageId: '',
          message: '',
        },
      });

      throw error;
    }
  }

  async clearQueue(actorId: string) {
    const queueUrl = configuration().sampleSqsQueueUrl;
    const queueId = queueUrl.split('/').pop() ?? queueUrl;

    try {
      const res = await this.sqsService.clearQueue(queueUrl);

      if (!res.ok) {
        throw new Error('Clearing SQS queue failed');
      }

      await this.databaseService.db.insert(domainEvents).values({
        eventType: 'clear_sample_sqs_queue',
        actorId,
        payload: { success: true, error: '', queueUrl, queueId },
      });

      return { success: true, message: 'Cleared sample SQS queue.' };
    } catch (error) {
      console.error('Clearing sample queue failed:', error);

      await this.databaseService.db.insert(domainEvents).values({
        eventType: 'clear_sample_sqs_queue',
        actorId,
        payload: {
          success: false,
          error: error instanceof Error ? error.message : String(error),
          queueUrl,
          queueId,
        },
      });

      throw error;
    }
  }
}

File: apps/newapi/services/sample_service.go
package services

import (
	"fmt"

	"gaeldle/newapi/config"
)

type SampleService struct {
	sqsService        *SqsService
	s3Service         *S3Service
	r2Service         *R2Service
	sampleSqsQueueURL string
}

func NewSampleService(sqsService *SqsService, s3Service *S3Service, r2Service *R2Service, cfg *config.AppConfig) *SampleService {
	return &SampleService{
		sqsService:        sqsService,
		s3Service:         s3Service,
		r2Service:         r2Service,
		sampleSqsQueueURL: cfg.SampleSqsQueueURL,
	}
}

func (s *SampleService) UploadImage(input interface{}, actorID string) (map[string]interface{}, error) {
	url := fmt.Sprintf("%s/test-dir/dummy-sample.jpg", s.r2Service.PublicURL)
	return map[string]interface{}{
		"success": true,
		"url":     url,
	}, nil
}

func (s *SampleService) SendMessage(input interface{}, actorID string) (map[string]interface{}, error) {
	if s.sampleSqsQueueURL == "" {
		return map[string]interface{}{
			"success":   true,
			"messageId": "dummy-sqs-message-id",
			"message":   "Message successfully queued",
		}, nil
	}

	msgID, err := s.sqsService.SendMessage(s.sampleSqsQueueURL, input)
	if err != nil {
		return nil, err
	}

	return map[string]interface{}{
		"success":   true,
		"messageId": msgID,
		"message":   "Message successfully queued",
	}, nil
}

func (s *SampleService) ClearQueue(actorID string) (map[string]interface{}, error) {
	if s.sampleSqsQueueURL == "" {
		return map[string]interface{}{
			"success": true,
			"message": "Queue successfully cleared",
		}, nil
	}

	err := s.sqsService.ClearQueue(s.sampleSqsQueueURL)
	if err != nil {
		return nil, err
	}

	return map[string]interface{}{
		"success": true,
		"message": "Queue successfully cleared",
	}, nil
}
